File: Go/include/gui.go
package include

import (
	"fmt"
	"strings"
)

type GUI struct {
	grid [22][22]uint8
}

func NewGUI() *GUI {
	gui := &GUI{}

	for i := range gui.grid {
		for j := range gui.grid[i] {
			gui.grid[i][j] = ' '
		}
	}

	return gui
}

func PlaceBorders(gui *GUI) {
	for i := range gui.grid {
		gui.grid[0][i] = '#'
		gui.grid[21][i] = '#'
		gui.grid[i][0] = '#'
		gui.grid[i][21] = '#'
	}
}

func PlaceSnake(gui *GUI, snake *Snake) {
	if snake == nil || snake.Head == nil {
		return
	}

	curr := snake.Head

	gui.grid[curr.X+1][curr.Y+1] = 'X'

	curr = curr.Next

	for ; curr != nil; curr = curr.Next {
		gui.grid[curr.X+1][curr.Y+1] = 'O'
	}
}

func PlaceApple(gui *GUI, apple *Apple) {
	gui.grid[apple.X+1][apple.Y+1] = '@'
}

func GenGUI(gui *GUI, snake *Snake, apple *Apple, score int) {
	var out strings.Builder

	GotoXY(1, 1)

	for i := range gui.grid {
		for j := range gui.grid[i] {
			gui.grid[i][j] = ' '
		}
	}

	PlaceBorders(gui)
	PlaceSnake(gui, snake)
	PlaceApple(gui, apple)

	for j := range 22 {
		for i := range 22 {
			field := gui.grid[i][j]

			if field == '#' {
				out.WriteString("##")
			} else {
				out.WriteByte(field)
				out.WriteByte(' ')
			}
		}

		if j == 0 {
			out.WriteString(fmt.Sprintf("  Score:  %d", score))
		}
		out.WriteByte('\n')
	}
	fmt.Print(out.String())
}

func GameOver() {
	GotoXY(1, 23)

	fmt.Println("         ___                 ")
    fmt.Println("        / __|__ _ _ __  ___  ")
    fmt.Println("       | (_ / _` | ''  \\/ -_) ")
    fmt.Println("        \\___\\__,_|_|_|_\\___| ")
    fmt.Println("         ___                 ")
    fmt.Println("        / _ \\_ _____ _ _     ")
    fmt.Println("       | (_) \\ V / -_) ''_|   ")
    fmt.Println("        \\___/ \\_/\\___|_|     ")
}

File: CSharp/include/GUI.cs
using System.Runtime.InteropServices;

namespace Ouroboros.Include;

class GUI
{
    private readonly char[,] Map;

    public GUI()
    {
        Map = new char[22, 22];
        MemoryMarshal.CreateSpan(ref Map[0, 0], Map.Length).Fill(' ');
    }

    public void GenGUI(Snake snake, Apple apple, int score)
    {
        Terminal.GotoXY(1, 1);

        MemoryMarshal.CreateSpan(ref Map[0, 0], Map.Length).Fill(' ');

        PlaceBorders();
        PlaceSnake(snake);
        PlaceApple(apple);

        for (int j = 0; j < 22; ++j)
        {
            for (int i = 0; i < 22; ++i)
            {
                var field = Map[i, j];

                if (field == '#')
                {
                    Console.Out.Write("##");
                }
                else
                {
                    Console.Out.Write($"{field} ");
                }
            }

            if (j == 0)
            {
                Console.Out.Write($"  Score:  {score}");
            }

            Console.Out.WriteLine();
        }
    }

    public static void GameOver()
    {
        Terminal.GotoXY(1, 23);

        Console.Out.WriteLine("         ___                 ");
        Console.Out.WriteLine("        / __|__ _ _ __  ___  ");
        Console.Out.WriteLine("       | (_ / _` | ''  \\/ -_) ");
        Console.Out.WriteLine("        \\___\\__,_|_|_|_\\___| ");
        Console.Out.WriteLine("         ___                 ");
        Console.Out.WriteLine("        / _ \\_ _____ _ _     ");
        Console.Out.WriteLine("       | (_) \\ V / -_) ''_|   ");
        Console.Out.WriteLine("        \\___/ \\_/\\___|_|     ");
    }

    private void PlaceBorders()
    {
        int height = Map.GetLength(0);
        int width = Map.GetLength(1);

        MemoryMarshal.CreateSpan(ref Map[0, 0], width).Fill('#');

        MemoryMarshal.CreateSpan(ref Map[height - 1, 0], width).Fill('#');

        for (int i = 0; i < height; ++i)
        {
            Map[i, 0] = '#';
            Map[i, width - 1] = '#';
        }
    }

    private void PlaceSnake(Snake snake)
    {
        bool first = true;
        foreach (var node in snake.GetNodes())
        {
            if (first)
            {
                Map[node.X + 1, node.Y + 1] = 'X';
                first = false;
            }
            else
            {
                Map[node.X + 1, node.Y + 1] = 'O';
            }
        }
    }

    private void PlaceApple(Apple apple)
    {
        Map[apple.X + 1, apple.Y + 1] = '@';
    }
}